runtime/io/common/src/aws/smithy/kotlin/runtime/io/SdkBufferedSink.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io
+
+/**
+ * A sink that keeps a buffer internally so that callers can do small writes without
+ * a performance penalty.
+ */
+public expect sealed interface SdkBufferedSink : SdkSink {
+    /**
+     * The underlying buffer for this sink
+     */
+    public val buffer: SdkBuffer
+
+    /**
+     * Write [limit] bytes from [source] starting at [offset]
+     */
+    public fun write(source: ByteArray, offset: Int = 0, limit: Int = source.size - offset): Unit
+
+    /**
+     * Write all bytes from [source] to this sink.
+     * @return the number of bytes read which will be 0 if [source] is exhausted
+     */
+    public fun writeAll(source: SdkSource): Long
+
+    /**
+     * Removes [byteCount] bytes from [source] and writes them to this sink.
+     */
+    public fun write(source: SdkSource, byteCount: Long): Unit
```

Comment:
The `SdkSink`, `SdkSource`, `SdkBufferedSink`, and `SdkBufferedSource` all basically mirror `okio` equivalents without any regard for making different subjective choices in API design. This is basically the answer to all of your "why" questions I think.

There are two reasons for mirroring (1) the API design in okio is reasonable (even if it has some additional overloads that could be condensend into one) and (2) (more importantly) having 1:1 interfaces makes it easy to go back and forth between the two. That being said if there are some that you think we should/could change while retaining the ability to convert between the two then we can talk (a few of your suggestions definitely fall into that category I think)

Reply:
> (1) makes less sense to me since we decided we didn't want to directly use Okio's types (because we can't on JS and because 3P API dependencies are not ideal). I think if we're going to make our own façade/version for these, then we should aim to make them as ergonomic and idiomatic as the rest of our SDK.

I was merely saying the API design choices they have made are reasonable and I didn't see anything glaringly bad that needed corrected. That being said if there are specific APIs that you think we should change we can do that.

> Are there particular changes I suggested that would make it difficult/impossible to adapt to/from Okio? At least some of them seem like simple name/overload changes but maybe there's more complexity behind others.

In the case of `write` vs `writeAll` they are separate because a source doesn't have a `size` property. The string/utf8 naming we can change. Were there any others other than that?

Reply:
> source doesn't have a `size` property

Oh right, well I guess ignore this particular suggestion then.